apps/api/src/app/partner-integrations/usecases/setup-vercel-integration/setup-vercel-integration.usecase.ts

```diff
@@ -0,0 +1,155 @@
+import { HttpService } from '@nestjs/axios';
+import { Injectable, NotFoundException } from '@nestjs/common';
+import { lastValueFrom } from 'rxjs';
+import { GetApiKeysCommand } from '../../../environments/usecases/get-api-keys/get-api-keys.command';
+import { GetApiKeys } from '../../../environments/usecases/get-api-keys/get-api-keys.usecase';
+import { GetEnvironment, GetEnvironmentCommand } from '../../../environments/usecases/get-environment';
+import { ApiException } from '../../../shared/exceptions/api.exception';
+import { SetupVercelIntegrationResponseDto } from '../../dtos/setup-vercel-integration-response.dto';
+import { SetupVercelIntegrationCommand } from './setup-vercel-integration.command';
+
+interface ISetEnvironment {
+  token: string;
+  projectIds: string[];
+  teamId: string | null;
+  identifier: string;
+  secretKey: string;
+}
+
+@Injectable()
+export class SetupVercelIntegration {
+  constructor(
+    private httpService: HttpService,
+    private getApiKeyUsecase: GetApiKeys,
+    private getEnvironmentUsecase: GetEnvironment
+  ) {}
+
+  async execute(command: SetupVercelIntegrationCommand): Promise<SetupVercelIntegrationResponseDto> {
+    try {
+      const tokenData = await this.getVercelToken(command.vercelIntegrationCode);
+
+      const projects = await this.getVercelProjects(tokenData.accessToken, tokenData.teamId);
+
+      const projectIds = projects.map((project) => project.id);
+
+      const envKeys = await this.getEnvKeys(command);
+
+      await this.setEnvironments({
+        projectIds,
+        identifier: envKeys.clientKey,
+        secretKey: envKeys.privateKey,
+        teamId: tokenData.teamId,
+        token: tokenData.accessToken,
+      });
+
+      return {
+        success: true,
+      };
+    } catch (error) {
+      throw new ApiException(
+        error?.response?.data?.error_description || error?.response?.data?.message || error.message
+      );
+    }
+  }
+
+  async getVercelToken(code: string): Promise<{
+    accessToken: string;
+    userId: string;
+    teamId: string | null;
+  }> {
+    const postData = new URLSearchParams({
+      code,
+      client_id: process.env.VERCEL_CLIENT_ID,
+      client_secret: process.env.VERCEL_CLIENT_SECRET,
+      redirect_uri: process.env.VERCEL_REDIRECT_URI,
+    });
+
+    const response = await lastValueFrom(
```

Comment:
I have never used `RxJS` library in the past. What are the benefits of using it for third party HTTP calls?
I just have a small concern about the warning in the usage of `lastValueFrom` stated here: https://rxjs.dev/api/index/function/lastValueFrom

> WARNING: Only use this with observables you know will complete. If the source observable does not complete, you will end up with a promise that is hung up, and potentially all of the state of an async function hanging out in memory. To avoid this situation, look into adding something like timeout, take, takeWhile, or takeUntil amongst others.

Reply:
the `httpModule` of nestjs doesn't returns a promise, rather it returns an Observable<AxiosResponse>
to convert it to promise, the lastValueFrom function is needed, 
maybe then we can use the axios directly?
https://docs.nestjs.com/techniques/http-module#using-axios-directly

Reply:
Thanks for the explanation. I am happy with this implementation anyway. Maybe it should be a topic to discuss with the rest of the core. 🙂